core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java

```diff
@@ -0,0 +1,167 @@
+                                .put(indexSettings())
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(2, 7))
+                                .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                                .build();
+        final int numNodes = randomIntBetween(1, numReplicas + 1);
+        internalCluster().ensureAtLeastNumDataNodes(numNodes);
+        assert internalCluster().numDataNodes() >= numNodes;
+        CreateIndexResponse response = prepareCreate(indexName)
+                                           .setSettings(settings)
+                                           .setWaitForActiveShards(ActiveShardCount.from(randomIntBetween(1, numNodes)))
+                                           .setTimeout("2s")
+                                           .get();
+        assertTrue(response.isAcknowledged());
+    }
+
+    public void testCreateIndexWaitsForAllActiveShards() throws Exception {
+        final String indexName = "test-idx";
+        // not enough data nodes, index creation times out
+        final int numReplicas = randomIntBetween(1, 3);
+        Settings settings = Settings.builder()
+                                .put(indexSettings())
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(2, 5))
+                                .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                                .build();
+        internalCluster().ensureAtMostNumDataNodes(numReplicas);
+        assertThrows(prepareCreate(indexName)
+                         .setSettings(settings)
+                         .setWaitForActiveShards(ActiveShardCount.ALL)
+                         .setTimeout("2s"), ElasticsearchTimeoutException.class);
+        if (client().admin().indices().prepareExists(indexName).get().isExists()) {
+            assertAcked(client().admin().indices().prepareDelete(indexName));
+        }
+
+        // enough data nodes, all shards are active
+        internalCluster().ensureAtLeastNumDataNodes(numReplicas + 1);
+        assert internalCluster().numDataNodes() >= numReplicas + 1;
+        settings = Settings.builder()
+                       .put(indexSettings())
+                       .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(2, 7))
+                       .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                       .build();
+        CreateIndexRequestBuilder request = prepareCreate(indexName)
+                      .setSettings(settings)
+                      .setWaitForActiveShards(ActiveShardCount.ALL)
+                      .setTimeout("2s");
+        assertTrue(request.get().isAcknowledged());
+    }
+
+    public void testInvalidWaitsValues() {
+        final String indexName = "test-idx";
+        try {
+            prepareCreate(indexName).setWaitForActiveShards(ActiveShardCount.from(-3));
```

Comment:
this actually only tests `ActiveShardCount.from(-3)` and should be a unit test instead.